File: src/GameView.java
/*
 * Copyright 2013 Mark Injerd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import java.awt.Dimension;
import java.awt.Font;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Image;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.awt.font.TextLayout;
import java.awt.geom.Rectangle2D;
import java.io.IOException;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;

import javax.imageio.ImageIO;
import javax.swing.Box;
import javax.swing.JButton;
import javax.swing.JComponent;
import javax.swing.JMenuBar;
import javax.swing.JMenuItem;
import javax.swing.JOptionPane;
import javax.swing.JPopupMenu;

public class GameView extends JComponent implements KeyListener {
	private static final long serialVersionUID = 1L;
	
	static final int VIEW_WIDTH = 1024, VIEW_HEIGHT = 768;
	private Image imgBg;
	
	/**
	 * Creates the game view and objects required within it.
	 * @throws IOException if any images could not be read.
	 */
	GameView() throws IOException {
		setPreferredSize(new Dimension(VIEW_WIDTH, VIEW_HEIGHT));
		imgBg = ImageIO.read(getClass().getClassLoader().getResource("img/background.png"));
		Audio.init();
		Bullet.init();
		LevelManager.createBackgroundAsteroids();
		addKeyListener(this);
		setFocusable(true);
		new Timer().schedule(new TimerTask() {
			@Override
			public void run() {
				if (!Simulation.isPaused()) {
					Simulation.simulate();
					repaint();
				}
			}
		}, 1000 / Simulation.TICK_RATE, 1000 / Simulation.TICK_RATE);
	}
	
	@Override
	protected void paintComponent(Graphics g) {
		super.paintComponent(g);
		Graphics2D g2d = (Graphics2D)g;
		g2d.drawImage(imgBg, 0, 0, null);
		g2d.setFont(new Font("Arial", Font.PLAIN, 22));
		List<Ship> ships = ShipManager.getShips();
		for (int i = 0; i < ships.size(); i++) {
			int x = 0, y = 0;
			if (i == 0 || i == 3) {
				x = 30;
			} else {
				x = VIEW_WIDTH - 118;
			}
			if (i == 0 || i == 2) {
				y = 30;
			} else {
				y = VIEW_HEIGHT - 42;
			}
			g2d.setColor(RenderUtils.PLAYER_COLORS[i]);
			g2d.drawString(String.format("%07d", ships.get(i).getScore()), x, y);
			RenderUtils.drawLives(g2d, ships.get(i).getLives(), RenderUtils.PLAYER_COLORS[i], x, y + 14);
		}
		Asteroid.drawAsteroids(g2d);
		for (Ship s : ships) {
			s.drawShip(g2d);
			Bullet.drawBullets(g2d, s);
		}
		if (LevelManager.isWaitingToStartLevel()) {
			Font font = new Font(Font.SANS_SERIF, Font.BOLD, 100);
			g2d.setFont(font);
			g2d.setColor(RenderUtils.TEXT_LEVEL_COLOR);
			String text = "LEVEL " + LevelManager.getLevel();
			Rectangle2D textBounds = new TextLayout(text, font, g2d.getFontRenderContext()).getBounds();
			g2d.drawString(text, VIEW_WIDTH / 2 - (int)textBounds.getWidth() / 2, VIEW_HEIGHT / 4 + (int)textBounds.getHeight() / 2);
		} else if (LevelManager.shouldShowText() && LevelManager.isGameOver()) {
			Font font = new Font(Font.SANS_SERIF, Font.BOLD, 100);
			g2d.setFont(font);
			g2d.setColor(RenderUtils.TEXT_GAMEOVER_COLOR);
			String text = "GAME OVER";
			Rectangle2D textBounds = new TextLayout(text, font, g2d.getFontRenderContext()).getBounds();
			g2d.drawString(text, VIEW_WIDTH / 2 - (int)textBounds.getWidth() / 2, VIEW_HEIGHT / 2 + (int)textBounds.getHeight() / 2);
		}
		getToolkit().sync();
	}
	
	JMenuBar createMenu() {
		JMenuBar menuBar = new JMenuBar();
		JButton button1 = new JButton("1");
		JButton button2 = new JButton("2");
		JButton button3 = new JButton("3");
		JButton button4 = new JButton("4");
		JButton buttonStart = new JButton("Start");
		JButton buttonPlayers = new JButton("Players");
		JButton buttonGame = new JButton("Game");
		JPopupMenu popupGame = new JPopupMenu();
		final JMenuItem menuHost = new JMenuItem("Host", KeyEvent.VK_H);
		final JMenuItem menuConnect = new JMenuItem("Connect", KeyEvent.VK_N);
		final JMenuItem menuDisconnect = new JMenuItem("Disconnect", KeyEvent.VK_D);
		JButton buttonPause = new JButton("Pause");
		JButton buttonSound = new JButton("Sound");
		JButton buttonMusic = new JButton("Music");
		JButton buttonHelp = new JButton("Help");
		JButton buttonAbout = new JButton("About");
		menuDisconnect.setVisible(false);
		menuBar.setFocusable(false);
		button1.setFocusable(false);
		button2.setFocusable(false);
		button3.setFocusable(false);
		button4.setFocusable(false);
		buttonStart.setFocusable(false);
		buttonPlayers.setFocusable(false);
		buttonGame.setFocusable(false);
		buttonPause.setFocusable(false);
		buttonSound.setFocusable(false);
		buttonMusic.setFocusable(false);
		buttonHelp.setFocusable(false);
		buttonAbout.setFocusable(false);
		button1.setMnemonic(KeyEvent.VK_1);
		button2.setMnemonic(KeyEvent.VK_2);
		button3.setMnemonic(KeyEvent.VK_3);
		button4.setMnemonic(KeyEvent.VK_4);
		buttonStart.setMnemonic(KeyEvent.VK_T);
		buttonPlayers.setMnemonic(KeyEvent.VK_L);
		buttonGame.setMnemonic(KeyEvent.VK_G);
		buttonPause.setMnemonic(KeyEvent.VK_P);
		buttonSound.setMnemonic(KeyEvent.VK_S);
		buttonMusic.setMnemonic(KeyEvent.VK_M);
		buttonHelp.setMnemonic(KeyEvent.VK_H);
		buttonAbout.setMnemonic(KeyEvent.VK_A);
		buttonStart.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				if (Simulation.isStarted()) {
					LevelManager.stopGame();
				} else {
					LevelManager.startGame();
				}
			}
		});
		buttonGame.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				popupGame.show(buttonGame, 0, buttonGame.getBounds().height);
			}
		});
		menuHost.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				Simulation.setStarted(false);
				MultiplayerManager.getInstance().startHost();
			}
		});
		menuConnect.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				String address = (String)JOptionPane.showInputDialog(GameView.this, "Address", "Connect", JOptionPane.QUESTION_MESSAGE);
				if (address != null) {
					MultiplayerManager.getInstance().connect(address);
				}
			}
		});
		menuDisconnect.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				MultiplayerManager.getInstance().disconnect();
			}
		});
		buttonPause.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				Simulation.setPaused(!Simulation.isPaused());
			}
		});
		buttonSound.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				buttonSound.setText(Audio.toggleSound() ? "Sound" : "(sound)");
			}
		});
		buttonMusic.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				buttonMusic.setText(Audio.toggleMusic() ? "Music" : "(music)");
				if (Simulation.isStarted()) {
					Audio.MUSIC_GAME.loop();
				}
			}
		});
		popupGame.add(menuHost);
		popupGame.add(menuConnect);
		popupGame.add(menuDisconnect);
		menuBar.add(button1);
		menuBar.add(button2);
		menuBar.add(button3);
		menuBar.add(button4);
		menuBar.add(Box.createHorizontalStrut(10));
		menuBar.add(buttonStart);
		menuBar.add(buttonPlayers);
		menuBar.add(buttonGame);
		menuBar.add(buttonPause);
		menuBar.add(buttonSound);
		menuBar.add(buttonMusic);
		menuBar.add(buttonHelp);
		menuBar.add(buttonAbout);
		Simulation.addGameStateListener(new Simulation.GameStateListener() {
			@Override
			public void onGameStartStateChanged(boolean started) {
				buttonStart.setText(started ? "Stop" : "Start");
				if (MultiplayerManager.getInstance().isClient()) {
					buttonStart.setEnabled(false);
					if (started) {
						LevelManager.startGame();
					} else if (!LevelManager.isGameOver()) {
						LevelManager.stopGame();
					}
				}
			}

			@Override
			public void onGamePauseStateChanged(boolean paused) {
				buttonPause.setText(paused ? "Continue" : "Pause");
				buttonPause.setMnemonic(paused ? KeyEvent.VK_C : KeyEvent.VK_P);
				if (MultiplayerManager.getInstance().isClient()) {
					Simulation.setPaused(paused);
				}
			}
		});
		MultiplayerManager.getInstance().setConnectionStateListener(new MultiplayerManager.ConnectionStateListener() {
			@Override
			public void onHostWaiting() {
				menuHost.setVisible(false);
				menuConnect.setVisible(false);
				menuDisconnect.setVisible(true);
			}
			@Override
			public void onConnected() {
				menuHost.setVisible(false);
				menuConnect.setVisible(false);
				menuDisconnect.setVisible(true);
				if (MultiplayerManager.getInstance().isClient()) {
					buttonStart.setEnabled(false);
					buttonPause.setEnabled(false);
				}
			}

			@Override
			public void onDisconnected() {
				menuHost.setVisible(true);
				menuConnect.setVisible(true);
				menuDisconnect.setVisible(false);
				buttonStart.setEnabled(true);
				buttonPause.setEnabled(true);
			}
		});
		return menuBar;
	}

	@Override
	public void keyPressed(KeyEvent e) {
		Ship ship = ShipManager.getLocalShip();
		if (e.isConsumed() || ship.isDestroyed() || Simulation.isPaused()) return;
		switch (e.getKeyCode()) {
			case KeyEvent.VK_UP:
				ship.thrust(true);
				break;
			case KeyEvent.VK_LEFT:
				ship.rotateLeft();
				break;
			case KeyEvent.VK_RIGHT:
				ship.rotateRight();
				break;
			case KeyEvent.VK_CONTROL:
				ship.fire();
				break;
		}
	}
	
	@Override
	public void keyReleased(KeyEvent e) {
		if (e.isConsumed()) return;
		Ship ship = ShipManager.getLocalShip();
		switch (e.getKeyCode()) {
			case KeyEvent.VK_UP:
				ship.thrust(false);
				break;
			case KeyEvent.VK_LEFT:
			case KeyEvent.VK_RIGHT:
				ship.rotateStop();
				break;
		}
	}
	
	@Override
	public void keyTyped(KeyEvent e) {}
}

File: src/GameView.kt
/*
 * Copyright 2013 Mark Injerd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import MultiplayerManager.ConnectionStateListener
import Simulation.GameStateListener
import java.awt.*
import java.awt.event.KeyEvent
import java.awt.event.KeyListener
import java.awt.font.TextLayout
import java.io.IOException
import java.util.*
import java.util.Timer
import javax.imageio.ImageIO
import javax.swing.*

/**
 * Creates the game view and objects required within it.
 * @throws IOException if any images could not be read.
 */
class GameView @Throws(IOException::class) internal constructor() : JComponent(), KeyListener {
	private val imgBg: Image

	init {
		preferredSize = Dimension(VIEW_WIDTH, VIEW_HEIGHT)
		imgBg = ImageIO.read(javaClass.getResource("background.png"))
		Audio.init()
		Bullet.init()
		LevelManager.createBackgroundAsteroids()
		addKeyListener(this)
		isFocusable = true
		Timer().schedule(object : TimerTask() {
			override fun run() {
				if (!Simulation.isPaused()) {
					Simulation.simulate()
					repaint()
				}
			}
		}, (1000 / Simulation.TICK_RATE).toLong(), (1000 / Simulation.TICK_RATE).toLong())
	}

	override fun paintComponent(g: Graphics) {
		super.paintComponent(g)
		val g2d = g as Graphics2D
		g2d.drawImage(imgBg, 0, 0, null)
		g2d.font = Font("Arial", Font.PLAIN, 22)
		val ships = ShipManager.ships
		synchronized(ships) {
			for ((i, ship) in ships.withIndex()) {
				ship ?: continue
				val x = if (i == 0 || i == 3) 30 else VIEW_WIDTH - 118
				val y = if (i == 0 || i == 2) 30 else VIEW_HEIGHT - 42
				g2d.color = RenderUtils.PLAYER_COLORS[i]
				g2d.drawString(String.format("%07d", ship.score), x, y)
				RenderUtils.drawLives(g2d, ship.lives, RenderUtils.PLAYER_COLORS[i], x, y + 14)
			}
		}
		Asteroid.drawAsteroids(g2d)
		synchronized(ships) {
			for (s in ships) {
				if (s == null) continue
				s.drawShip(g2d)
				Bullet.drawBullets(g2d, s)
			}
		}
		if (LevelManager.isWaitingToStartLevel) {
			val font = Font(Font.SANS_SERIF, Font.BOLD, 100)
			g2d.font = font
			g2d.color = RenderUtils.TEXT_LEVEL_COLOR
			val text = "LEVEL ${LevelManager.level}"
			val textBounds = TextLayout(text, font, g2d.fontRenderContext).bounds
			g2d.drawString(text, VIEW_WIDTH / 2 - textBounds.width.toInt() / 2, VIEW_HEIGHT / 4 + textBounds.height.toInt() / 2)
		} else if (LevelManager.shouldShowText() && LevelManager.isGameOver) {
			val font = Font(Font.SANS_SERIF, Font.BOLD, 100)
			g2d.font = font
			g2d.color = RenderUtils.TEXT_GAMEOVER_COLOR
			val text = "GAME OVER"
			val textBounds = TextLayout(text, font, g2d.fontRenderContext).bounds
			g2d.drawString(text, VIEW_WIDTH / 2 - textBounds.width.toInt() / 2, VIEW_HEIGHT / 2 + textBounds.height.toInt() / 2)
		}
		toolkit.sync()
	}

	fun createMenu(): JMenuBar {
		val menuBar = JMenuBar()
		val button1 = JButton("1")
		val button2 = JButton("2")
		val button3 = JButton("3")
		val button4 = JButton("4")
		val buttonStart = JButton("Start")
		val buttonPlayers = JButton("Players")
		val buttonGame = JButton("Game")
		val popupGame = JPopupMenu()
		val menuHost = JMenuItem("Host", KeyEvent.VK_H)
		val menuConnect = JMenuItem("Connect", KeyEvent.VK_N)
		val menuDisconnect = JMenuItem("Disconnect", KeyEvent.VK_D)
		val buttonPause = JButton("Pause")
		val buttonSound = JButton("Sound")
		val buttonMusic = JButton("Music")
		val buttonHelp = JButton("Help")
		val buttonAbout = JButton("About")
		menuDisconnect.isVisible = false
		menuBar.isFocusable = false
		button1.isFocusable = false
		button2.isFocusable = false
		button3.isFocusable = false
		button4.isFocusable = false
		buttonStart.isFocusable = false
		buttonPlayers.isFocusable = false
		buttonGame.isFocusable = false
		buttonPause.isFocusable = false
		buttonSound.isFocusable = false
		buttonMusic.isFocusable = false
		buttonHelp.isFocusable = false
		buttonAbout.isFocusable = false
		button1.mnemonic = KeyEvent.VK_1
		button2.mnemonic = KeyEvent.VK_2
		button3.mnemonic = KeyEvent.VK_3
		button4.mnemonic = KeyEvent.VK_4
		buttonStart.mnemonic = KeyEvent.VK_T
		buttonPlayers.mnemonic = KeyEvent.VK_L
		buttonGame.mnemonic = KeyEvent.VK_G
		buttonPause.mnemonic = KeyEvent.VK_P
		buttonSound.mnemonic = KeyEvent.VK_S
		buttonMusic.mnemonic = KeyEvent.VK_M
		buttonHelp.mnemonic = KeyEvent.VK_H
		buttonAbout.mnemonic = KeyEvent.VK_A
		button1.isEnabled = false
		button2.isEnabled = false
		button3.isEnabled = false
		button4.isEnabled = false
		buttonPlayers.isEnabled = false
		buttonHelp.isEnabled = false
		buttonAbout.isEnabled = false
		buttonStart.addActionListener {
			if (Simulation.isStarted) {
				LevelManager.stopGame()
			} else {
				LevelManager.startGame()
			}
		}
		buttonGame.addActionListener { popupGame.show(buttonGame, 0, buttonGame.bounds.height) }
		menuHost.addActionListener {
			Simulation.isStarted = false
			MultiplayerManager.instance.startHost()
		}
		menuConnect.addActionListener {
			val address = JOptionPane.showInputDialog(this@GameView, "Address", "Connect", JOptionPane.QUESTION_MESSAGE)
			if (address != null) {
				MultiplayerManager.instance.connect(address)
			}
		}
		menuDisconnect.addActionListener { MultiplayerManager.instance.disconnect() }
		buttonPause.addActionListener { Simulation.setPaused(!Simulation.isPaused()) }
		buttonSound.addActionListener { buttonSound.text = if (Audio.toggleSound()) "Sound" else "(sound)" }
		buttonMusic.addActionListener {
			buttonMusic.text = if (Audio.toggleMusic()) "Music" else "(music)"
			if (Simulation.isStarted) {
				Audio.MUSIC_GAME.loop()
			}
		}
		popupGame.add(menuHost)
		popupGame.add(menuConnect)
		popupGame.add(menuDisconnect)
		menuBar.add(button1)
		menuBar.add(button2)
		menuBar.add(button3)
		menuBar.add(button4)
		menuBar.add(Box.createHorizontalStrut(10))
		menuBar.add(buttonStart)
		menuBar.add(buttonPlayers)
		menuBar.add(buttonGame)
		menuBar.add(buttonPause)
		menuBar.add(buttonSound)
		menuBar.add(buttonMusic)
		menuBar.add(buttonHelp)
		menuBar.add(buttonAbout)
		Simulation.addGameStateListener(object : GameStateListener {
			override fun onGameStartStateChanged(started: Boolean) {
				buttonStart.text = if (started) "Stop" else "Start"
				if (MultiplayerManager.instance.isClient) {
					buttonStart.isEnabled = false
					if (started) {
						LevelManager.startGame()
					} else if (!LevelManager.isGameOver) {
						LevelManager.stopGame()
					}
				}
			}

			override fun onGamePauseStateChanged(paused: Boolean) {
				buttonPause.text = if (paused) "Continue" else "Pause"
				buttonPause.mnemonic = if (paused) KeyEvent.VK_C else KeyEvent.VK_P
				if (MultiplayerManager.instance.isClient) {
					Simulation.setPaused(paused)
				}
			}
		})
		MultiplayerManager.instance.setConnectionStateListener(object : ConnectionStateListener {
			override fun onHostWaiting() {
				menuHost.isVisible = false
				menuConnect.isVisible = false
				menuDisconnect.isVisible = true
			}

			override fun onConnected() {
				menuHost.isVisible = false
				menuConnect.isVisible = false
				menuDisconnect.isVisible = true
				if (MultiplayerManager.instance.isClient) {
					buttonStart.isEnabled = false
					buttonPause.isEnabled = false
				}
			}

			override fun onDisconnected() {
				menuHost.isVisible = true
				menuConnect.isVisible = true
				menuDisconnect.isVisible = false
				buttonStart.isEnabled = true
				buttonPause.isEnabled = true
			}
		})
		return menuBar
	}

	override fun keyPressed(e: KeyEvent) {
		val ship = ShipManager.localShip
		if (e.isConsumed || ship.isDestroyed || Simulation.isPaused()) return
		when (e.keyCode) {
			KeyEvent.VK_UP -> ship.thrust(true)
			KeyEvent.VK_LEFT -> ship.rotateLeft()
			KeyEvent.VK_RIGHT -> ship.rotateRight()
			KeyEvent.VK_SPACE, KeyEvent.VK_CONTROL -> ship.fire()
		}
	}

	override fun keyReleased(e: KeyEvent) {
		if (e.isConsumed) return
		val ship = ShipManager.localShip
		when (e.keyCode) {
			KeyEvent.VK_UP -> ship.thrust(false)
			KeyEvent.VK_LEFT, KeyEvent.VK_RIGHT -> ship.rotateStop()
		}
	}

	override fun keyTyped(e: KeyEvent) {}

	companion object {
		private const val serialVersionUID = 1L
		const val VIEW_WIDTH = 1024
		const val VIEW_HEIGHT = 768
	}
}
